File: lib/pbio/test/drv/button.c
// SPDX-License-Identifier: MIT
// Copyright (c) 2020 The Pybricks Authors

// Software button implementation for simulating buttons in tests

#include <pbio/button.h>
#include <pbio/error.h>

static pbio_button_flags_t pbio_test_button_flags;

void pbio_test_button_set_pressed(pbio_button_flags_t flags) {
    pbio_test_button_flags = flags;
}

void _pbdrv_button_init(void) {
}

pbio_error_t pbdrv_button_is_pressed(pbio_button_flags_t *pressed) {
    *pressed = pbio_test_button_flags;
    return PBIO_SUCCESS;
}

File: lib/pbio/cpython/pbio_virtual/drv/button.py
# SPDX-License-Identifier: MIT
# Copyright (c) 2022 The Pybricks Authors

from enum import IntFlag


class ButtonFlags(IntFlag):
    """
    Hub buttons.

    These values match ``pbio_button_flags_t``.
    """

    LEFT_DOWN = 1 << 1
    DOWN = 1 << 2
    RIGHT_DOWN = 1 << 3
    LEFT = 1 << 4
    CENTER = 1 << 5
    RIGHT = 1 << 6
    LEFT_UP = 1 << 7
    UP = 1 << 8
    RIGHT_UP = 1 << 9


class VirtualButtons:
    pressed: ButtonFlags = ButtonFlags(0)
    """
    Provides the button flags returned by ``pbdrv_button_is_pressed()``.

    CPython code should write to this attribute to simulate the buttons currently
    pressed and the PBIO driver will read this attribute.
    """
